pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java

```diff
@@ -1083,4 +1117,103 @@ private void createPendingLoadTopic() {
         }
 
     }
+    
+    /**
+     * Adds given dispatcher's unackMessage count to broker-unack message count and if it reaches to the
+     * {@link #maxUnackedMessages} then it blocks all the dispatchers which has unack-messages higher than
+     * {@link #maxUnackedMsgsPerDispatcher}. It unblocks all dispatchers once broker-unack message counts decreased to
+     * ({@link #maxUnackedMessages}/2)
+     * 
+     * @param dispatcher
+     * @param numberOfMessages
+     */
+    public void addUnAckedMessages(PersistentDispatcherMultipleConsumers dispatcher, int numberOfMessages) {
+        // don't block dispatchers if maxUnackedMessages = 0
+        if (maxUnackedMessages <= 0) {
+            return;
+        }
+        int unAckedMessages = TOTAL_UNACKED_MESSAGES_UPDATER.addAndGet(this, numberOfMessages);
```

Comment:
Also another issue is when [dispatcher calls this method to increment this number, broker checks if per-broker maxUnackMsg](https://github.com/rdhabalia/pulsar/blob/55dcaf507d8c51f3693df25adbafec16146075ea/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1141) limit reaches and then it tries to check that a caller dispatcher has reached to `maxPerDispatcherLimit (to block only misbehaving dispatchers)` and it blocks the dispatcher if it has reached the limit.

Now, if we schedule the task then we don't know the mis-behaving dispatchers so, we have to scan through all the list of topics' dispatchers to find out mis-behaved dispatcher and block them so, it won't impact other topics. So, do you think keeping it as AtomicInteger will be a critical point for contention?